File: countingHaybales.py
from typing import List, Tuple

class Solution:
    def countHaybales(self, haybales: List[int], queries: List[Tuple[int, int]]):
        n = len(haybales)
        Q = len(queries)

        haybales.sort()

        for query in queries:
            ub = self.upperBound(haybales, query[1])
            lb = self.lowerBound(haybales, query[0])

            print(ub - lb)

    def upperBound(self, haybales: List[int], target: int) -> int:
        low = 0
        high = len(haybales)

        while low < high:
            mid = low + (high - low) // 2

            if haybales[mid] <= target:
                low = mid + 1
            else:
                high = mid

        return low

    def lowerBound(self, haybales: List[int], target: int) -> int:
        low = 0
        high = len(haybales)

        while low < high:
            mid = low + (high - low) // 2

            if haybales[mid] < target:
                low = mid + 1
            else:
                high = mid

        return low

File: countinghaybales.java
import java.util.*;

class Solution {
    public void countHaybales(ArrayList<Integer> haybales, ArrayList<Pair<Integer, Integer>> queries) {
        int n = haybales.size();
        int Q = queries.size();

        Collections.sort(haybales);

        for (Pair<Integer, Integer> query : queries) {
            int ub = upperBound(haybales, query.second);
            int lb = lowerBound(haybales, query.first);

            System.out.println(ub - lb);
        }
    }

    private int upperBound(ArrayList<Integer> haybales, int target) {
        int low = 0;
        int high = haybales.size();

        while (low < high) {
            int mid = low + (high - low) / 2;

            if (haybales.get(mid) <= target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        return low;
    }

    private int lowerBound(ArrayList<Integer> haybales, int target) {
        int low = 0;
        int high = haybales.size();

        while (low < high) {
            int mid = low + (high - low) / 2;

            if (haybales.get(mid) < target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        return low;
    }
}
